### ISSUE TITLE
Validate SAML Response fail when Signature nodes exist in both <Response> and <Assertion>

### ISSUE THREAD

**Issue by Ailsa84** [tngan/samlify#127]
My SAML Response has Signature in both <Response> and <Assertion>, like following:

```xml
<samlp:Response xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" Destination="https://sp.com/auth/saml" ID="FIMRSP_a3219212-015e-1e38-826c-b87377b46274" IssueInstant="2017-09-21T06:30:39Z" Version="2.0">
  <saml:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">https://idp.sso.ibm.com/auth/sps/samlidp/saml20</saml:Issuer>
  <ds:Signature Id="uuida3219213-015e-1f4b-b1f4-b87377b46274">
    <ds:SignedInfo>
      <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
      <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
      <ds:Reference URI="#FIMRSP_a3219212-015e-1e38-826c-b87377b46274">
        <ds:Transforms>
          <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
          <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#">
            <xc14n:InclusiveNamespaces xmlns:xc14n="http://www.w3.org/2001/10/xml-exc-c14n#" PrefixList="samlp xs saml xsi ds"/>
          </ds:Transform>
        </ds:Transforms>
        <ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
        <ds:DigestValue>g0aZYrHvp6dqbuUtXiTJddxWqpL+lzMkJkD49mqxU1g=</ds:DigestValue>
      </ds:Reference>
    </ds:SignedInfo>
    <ds:SignatureValue>ZOLilj+Ws8LoBGjYRTFA2c02Pq8n2dH34MI4o+hMoANeuEWWbCrNI8a7jz0t5OyDMVsRXzXVDuW6bU8PJB3KlThBIs0cV3QWd1ecoY5m5ZrViJuKnPXivb4W7jx5aWFwrkb/O4EEYFESo89Kb+bZhC9Qai+a1R34zxezdor+Wy+6/TSLoi+N2rFCRAlo8gtqx3AnrI5jnFc74SIy+geALDKxR3RGPdG/roTFFpHkvkkYwqUUGnjgN8mZAm1P4aIkwqwgUl9/vjFroD2JkwGv3OD2rX/3UJ4vyYFkeFnEbIOaTvdF/AGbFjo4KAtfbYOPdVkctgi+GVKxyuv2kDWeJw==</ds:SignatureValue>
    <ds:KeyInfo>
      <ds:X509Data>
        <ds:X509Certificate>MIIDdzCCAl+gAwIBAgIENmZ8gDANBgkqhkiG9w0BAQsFADBsMQswCQYDVQQGEwJVUzELMAkGA1UECBMCTlkxDzANBgNVBAcTBkFybW9uazEQMA4GA1UEChMHaWJtLmNvbTEMMAoGA1UECxMDQ0lTMR8wHQYDVQQDExZ3M2lkLmFscGhhLnNzby5pYm0uY29tMB4XDTE1MTEyMDE3NDAyM1oXDTE4MTExOTE3NDAyM1owbDELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAk5ZMQ8wDQYDVQQHEwZBcm1vbmsxEDAOBgNVBAoTB2libS5jb20xDDAKBgNVBAsTA0NJUzEfMB0GA1UEAxMWdzNpZC5hbHBoYS5zc28uaWJtLmNvbTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAIMdVEidEhUKdfUFwR5+bOk12uJHtiE2D+6LyeMJOdLInWbNFrl3WP9TL5fjeF9oNACQcuQy7Ll2rt+UqUXbU3pVvT7bzoyfeJWAPjzRqFP2f/U40sBbftVHnbC8mRcwD7VyJqjB3IIfV78mYfrJUM6b3hYW2GGU9dzkNpKc1S7dDmbpkOwFqn4PgRVJNnS3eZq7T46hS6gjnfD1MAsDTPszkwMWB4jMQuhvswt3+fgcGBmKlQFVXidDRyg7aA8oYq5Mj07307ulywzt/oyXljQCai/9M1wg3MpuQ+Qb0iJ+W1oiBRpRUATbYuUFC3fPV886LmH7UOd6MxrWrGjuuS8CAwEAAaMhMB8wHQYDVR0OBBYEFAtr3Gj757xspsi5XIZI7E9041j+MA0GCSqGSIb3DQEBCwUAA4IBAQBGM7PALAwVEsqS8egOHMOGuKTSSPZVCdlxi8xpP4f5feoaK0f7gqiD2ujsi7Oy7IuZ4USAh61tjSmdayHDTwrfpih+vaMts5lz4O+Z5Ox/2itYggDOPp3ajohVfGn0EupWnRVy3RotiWQ9WvHeBTCitmEICLnSi4gYmfILeJKBCo5zrM1E2OD2ESdtTLwgQyx9xxL78Zf/yF0l8e4o1ttwuxx/e/IghC1OQcX2mIlH/68NABTYD3jLdNUK65zrbMhRPhWpqXEdOOmKdQjtcZPpihVawDSHGjVhqPaYam7/aAMJ5jJHuy2qZCM8ZgY05KVmlcdO/S4rCgeXLS8uMhpE</ds:X509Certificate>
      </ds:X509Data>
    </ds:KeyInfo>
  </ds:Signature>
  <samlp:Status>
    <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
  </samlp:Status>
  <saml:Assertion ID="Assertion-uuida3219202-015e-1f6a-8dd6-b87377b46274" IssueInstant="2017-09-21T06:30:39Z" Version="2.0">
    <saml:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">https://idp.sso.ibm.com/auth/sps/samlidp/saml20</saml:Issuer>
    <ds:Signature Id="uuida3219203-015e-107d-a444-b87377b46274">
      <ds:SignedInfo>
        <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
        <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
        <ds:Reference URI="#Assertion-uuida3219202-015e-1f6a-8dd6-b87377b46274">
          <ds:Transforms>
            <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
            <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#">
              <xc14n:InclusiveNamespaces xmlns:xc14n="http://www.w3.org/2001/10/xml-exc-c14n#" PrefixList="xs saml xsi"/>
            </ds:Transform>
          </ds:Transforms>
          <ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
          <ds:DigestValue>L22FDQTnvc0+ZboffCFebWKwxMdDqVovSrf2Og/XYh8=</ds:DigestValue>
        </ds:Reference>
      </ds:SignedInfo>
      <ds:SignatureValue>aiE/uHkS3M7ck74yaNsyLvGrcZED3JeTi3rttjftIFebcbc7cY+E1c8cctskONf8grZDPKfAWfhFlDMBuq/oyzoMjbcXuuM/KmksLXkzCd6bSM/P6BqW+LKC1...
```

When validate first DigestValue in <samlp:Response>, I found all signature nodes are replaced with '',
xml = xml.replace(/<ds:Signature(.*?)>(.*?)<\/(.*?)ds:Signature>/g, '');

But this will also replace signagure node in <saml:Assertion>, then send this 'xml' to 'xml-crypto' will calculate a different incorrect digestValue, so is it a bug or wrong options I set? Thanks.

Following is my options set:

```js
const idp = saml.IdentityProvider({
   metadata: fs.readFileSync('./samlidp_metadata.xml'),
       isAssertionEncrypted: false,
       wantMessageSigned: true,
       signatureConfig: {
            prefix: 'ds',
            location: {
              reference: '/samlp:Response/saml:Issuer',
              action: 'after',
            },
          },
});

const sp = saml.ServiceProvider({
  metadata: fs.readFileSync('./SP_meta.xml'),
  wantMessageSigned: true,
  wantAssertionsSigned: true,
});

sp
.parseLoginResponse(idp, 'post', { body: params })
.then((result) => {
  console.log('result=', result);
});
```

**Comment by tngan** [comment]
@Ailsa84 Sorry for the late reply. Will take a look soon.

**Comment by tngan** [comment]
This test case should cover the use case, parsing response with signed message and signed assertion.

https://github.com/tngan/samlify/blob/5537489b83320e1d99bd116e4bf761c148692d30/test/flow.ts#L341-L357

FYI: In `xml-crypto`, the signature verification doesn't require the document includes signature. You can check out the note section [here](https://github.com/yaronn/xml-crypto#verifying-xml-documents).

_The xml-crypto api requires you to supply it separately the xml signature ("<Signature>...</Signature>", in loadSignature) and the signed xml (in checkSignature). The signed xml may or may not contain the signature in it, but you are still required to supply the signature separately._

**Comment by jacqt** [comment]
I think I am currently running into a similar problem

After commenting out this line https://github.com/tngan/samlify/blob/master/src/libsaml.ts#L469 I was able to get samlify to validate the SAML response.

What was the reason behind this line?

@tngan

**Comment by tngan** [comment]
@jacqt Please try again with the latest samlify rc version, we have tackled a potential wrapping attack of response, and make the flow much more cleaner for debugging.

**Comment by tngan** [comment]
@jacqt @Ailsa84 These should be resolved finally via 4afbb10. Feel free to take a look for the code tracing detail in #219.
